### PR TITLE
Docstrings, Part 1

### PR DESCRIPTION
**Pull request by ffuuugor**
Docstring PRs promise to be quite lengthy, so for the reviewer's comfort I'll split them into multiple parts.

This episode covers all of the modules available at the top level: `privacy_engine.py`, `data_loader.py`,  `distributed.py` and `scheduler.py`

### CODE REVIEW

File: opacus/distributed.py [hunk 1]
```diff
@@ -15,7 +24,11 @@ def average_gradients(model):
 
 
 class DifferentiallyPrivateDistributedDataParallel(nn.Module):
-    def __init__(self, model):
+    """
+    Implements distributed data parallelism that is based on
+    ``torch.distributed`` package at the module level.
+    """
```
Comment (ashkan-software) [inline opacus/distributed.py:30]: you probably need a new line here

File: opacus/privacy_engine.py [hunk 2]
```diff
@@ -15,7 +15,20 @@
...
 
-def forbid_accumulation_hook(module: nn.Module, _):
+def forbid_accumulation_hook(module: GradSampleModule, _):
+    """
+    Model hook, that detects repetitive forward/backward passes between optimizer
+    steps.
+
+    Args:
+        module: input module
+        _:
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:25]: This one also need some explanation (description of a variable cannot be empty)

File: opacus/privacy_engine.py [hunk 3]
```diff
@@ -29,29 +42,43 @@ def forbid_accumulation_hook(module: nn.Module, _):
...
+        >>> privacy_engine = PrivacyEngine()
+        >>>
+        >>> model, optimizer, dataloader = privacy_engine.make_private(
+        ...    module=model,
+        ...    optimizer=optimizer,
+        ...    data_loader=dataloader,
+        ...    noise_multiplier=1.0,
+        ...    max_grad_norm=1.0,
+        ... )
+        >>> # continue training as normal
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:66]: Does this work with doctest? @romovpa

File: opacus/privacy_engine.py [hunk 4]
```diff
@@ -29,29 +42,43 @@ def forbid_accumulation_hook(module: nn.Module, _):
...
-            ...         optimizer.zero_grad()
-            ...     else:
-            ...         optimizer.virtual_step()  # this will call privacy engine's virtual_step()
 
+        Args:
+            accountant: Accounting mechanism. Currently supported:
+                - rdp (:class:`~opacus.accountants.RDPAccountant`)
+                - gdp (:class:`~opacus.accountants.GaussianAccountant`)
+            secure_mode: Set to ``True`` if cryptographically strong DP guarantee is
+                requires. ``secure_mode=True`` uses secure random number generator for
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:76]:
```suggestion
                required. ``secure_mode=True`` uses secure random number generator for
```

File: opacus/privacy_engine.py [hunk 5]
```diff
@@ -147,6 +174,14 @@ def is_compatible(
     ) -> bool:
         """
         Check if task components are compatible with DP.
+
+        Args:
+            module:
+            optimizer:
+            data_loader:
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:181]: please fill with description :)

File: opacus/privacy_engine.py [hunk 6]
```diff
@@ -159,6 +194,15 @@ def validate(
         """
         Validate that task components are compatible with DP.
         Same as ``is_compatible()``, but raises error instead of returning bool.
+
+        Args:
+            module:
+            optimizer:
+            data_loader:
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:201]: same here

File: opacus/privacy_engine.py [hunk 7]
```diff
@@ -167,6 +211,15 @@ def get_compatible_module(cls, module: nn.Module) -> nn.Module:
         """
         Return a privacy engine compatible module. Also validates the module after
         running registered fixes.
+
+        Args:
+            module:
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:216]: description please!

File: opacus/privacy_engine.py [hunk 8]
```diff
@@ -185,7 +238,59 @@ def make_private(
         noise_seed: Optional[int] = None,
         poisson_sampling: bool = True,
         clipping: str = "flat",
-    ):
+    ) -> Tuple[GradSampleModule, DPOptimizer, DataLoader]:
+        """
+        Add privacy-related responsibilites to the main PyTorch training objects:
+        model, optimizer and the data loader.
+
+        All of the returned objects act just like treir non-private counterparts
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:246]:
```suggestion
        All of the returned objects act just like their non-private counterparts
```

File: opacus/privacy_engine.py [hunk 9]
```diff
@@ -185,7 +238,59 @@ def make_private(
...
+        passed as arguments, but with added DP tasks.
+
+        Model is wrapped to also compute per sample gradients.
+        Optimizer is now responsible for gradient clipping and adding noise to the
+            gradients
+        DataLoader is updated to perform Poisson sampling.
+
+        Note, that using any other models, optimizers or data sources during training
+        will invalidate stated privacy guarantees.
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:255]: can put these 2 lines under `Notes:`

File: opacus/privacy_engine.py [hunk 10]
```diff
@@ -185,7 +238,59 @@ def make_private(
...
+            optimizer: Optimizer to be used for training
+            data_loader: DataLoader to be used for training
+            noise_multiplier: The ratio of the standard deviation of the Gaussian noise to
+                the L2-sensitivity of the function to which the noise is added
+                (How much noise to add)
+            max_grad_norm: The maximum norm of the per-sample gradients. Any gradient with norm
+                higher than this will be clipped to this value.
+            batch_first: Flag to indicate if the input tensor to the corresponding module
+                has the first dimension representing the batch. If set to True, dimensions on
+                input tensor will be ``[batch_size, ..., ...]``.
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:268]: should we mention if `False` the second argument is batch?

File: opacus/privacy_engine.py [hunk 11]
```diff
@@ -185,7 +238,59 @@ def make_private(
...
+                using Poisson sampling is unfeasible.
+            clipping: Per sample gradient clipping mechanism ("flat" or "per_layer").
+                Flat clipping calculates the norm of the entire gradient over
+                all parameters, while per layer clipping sets individual norms for
+                every parameter tensor. Flat clipping is usually preferred, but using
+                per layer clipping in combination with distributed training can provide
+                notable performance gains.
+
+        Returns:
+            Tuple of (model, optimizer, data_loader)
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:285]: This in the website will be something like this: "Tuple of (model, optimizer, data_loader)Model is a wrapper...". Perhaps can put a dot and space at the end of this line

File: opacus/privacy_engine.py [hunk 12]
```diff
@@ -267,4 +416,11 @@ def make_private_with_epsilon(
         )
 
     def get_epsilon(self, delta):
+        """
+        Computes the (epsilon, delta) privacy budget spent so far.
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:420]: need a new line after the method description

File: opacus/privacy_engine.py [hunk 13]
```diff
@@ -267,4 +416,11 @@ def make_private_with_epsilon(
...
+        Args:
+            target_delta: The Target delta.
+        Returns:
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:423]: need a new line after listing all the args

File: opacus/scheduler.py [hunk 14]
```diff
@@ -38,7 +38,18 @@ def step(self):
 
 
 class ExponentialNoise(_NoiseScheduler):
+    """
+    Decays the noise_multiplier by gamma every epoch.
+    When last_epoch=-1, sets initial noise_multiplier as noise_multiplier.
+    """
     def __init__(self, optimizer: DPOptimizer, gamma: float, last_epoch: int = -1):
```
Comment (ashkan-software) [inline opacus/scheduler.py:45]: need a new line before this

File: opacus/scheduler.py [hunk 15]
```diff
@@ -65,9 +88,21 @@ def get_noise_multiplier(self):
 
 
 class StepNoise(_NoiseScheduler):
+    """
+    Decays the noise_multiplier by gamma every step_size epochs.
+    When last_epoch=-1, sets initial noise_multiplier as noise_multiplier.
+    """
     def __init__(
```
Comment (ashkan-software) [inline opacus/scheduler.py:95]: new line before this

**Review by ashkan-software — CHANGES_REQUESTED** [review]
Great work. We can also follow this up internally or here.

File: conftest.py [hunk 16]
```diff
@@ -4,25 +4,26 @@
...
-    dataloader = []
-    for _ in range(64):
-        data = torch.randn(4, 16)
-        labels = torch.randint(0, 2, (4,))
-        dataloader.append((data, labels))
+    dataset = TensorDataset(
+        torch.randn(64, 5),
+        torch.randint(0, 2, (64,))
+    )
+    dataloader = DataLoader(dataset, batch_size=4)
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply conftest.py:26]: I don't trust myself all that much, but the new tests generated from this are passing :)

File: opacus/data_loader.py [hunk 17]
```diff
@@ -13,6 +13,18 @@
 def wrap_collate_with_empty(
     collate_fn: Optional[_collate_fn_t], sample_empty_shapes: Sequence
 ):
+    """
+    Wraps given collate function to handle empty batches.
+
+    Args:
+        collate_fn: collate function to wrap
+        sample_empty_shapes: expected shape for a batch of size 0
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/data_loader.py:21]: It's actually a sequence. For most datasets it'll be smth like `[data.shape, labels.shape]`

Clarified the docstring

File: opacus/privacy_engine.py (hunk 3 above)
Comment (ffuuugor) [reply opacus/privacy_engine.py:66]: Yes, you can see this is executed and passed in the unit test run: https://app.circleci.com/pipelines/github/pytorch/opacus/1324/workflows/be756a10-42ec-43e0-ac32-999d13bb06fa/jobs/6942

```
opacus/per_sample_gradient_clip.py s                                     [  0%]
opacus/privacy_engine.py .                                               [  1%]
opacus/accountants/analysis/rdp.py .                                     [  2%]
opacus/grad_sample/utils.py .                                            [  3%]
opacus/layers/dp_multihead_attention.py .                                [  3%]
opacus/layers/param_rename.py .                                          [  4%]
opacus/tests/accountants_test.py ...                                     [  7%]
<...>
```

**Comment by ffuuugor** [comment]
@ashkan-software Thanks a lot for the very thorough review!

File: opacus/data_loader.py [hunk 18]
```diff
@@ -13,6 +13,20 @@
 def wrap_collate_with_empty(
     collate_fn: Optional[_collate_fn_t], sample_empty_shapes: Sequence
 ):
+    """
+    Wraps given collate function to handle empty batches.
+
+    Args:
+        collate_fn: collate function to wrap
+        sample_empty_shapes: expected shape for a batch of size 0. Input is a sequence -
+            one for each tensors in the dataset
```
Comment (karthikprasad) [inline opacus/data_loader.py:22]:
```suggestion
            one for each tensor in the dataset
```

File: opacus/data_loader.py [hunk 19]
```diff
@@ -22,17 +36,49 @@ def collate(batch):
...
+    Args:
+        x: any object
+
+    Returns:
+        ``x.shape`` if attribute exists, empty tuple otherwise
+    """
     if hasattr(x, "shape"):
         return x.shape
     else:
         return ()
```
Comment (karthikprasad) [inline opacus/data_loader.py:52]:
```suggestion
    return x.shape if hasattr(x, "shape") else ()
```

File: opacus/privacy_engine.py [hunk 20]
```diff
@@ -28,28 +42,42 @@ def forbid_accumulation_hook(module: GradSampleModule, _):
...
 
+        Args:
+            accountant: Accounting mechanism. Currently supported:
+                - rdp (:class:`~opacus.accountants.RDPAccountant`)
+                - gdp (:class:`~opacus.accountants.GaussianAccountant`)
+            secure_mode: Set to ``True`` if cryptographically strong DP guarantee is
+                required. ``secure_mode=True`` uses secure random number generator for
+                noise and shuffling (as opposed to pseudo-rng in vanilla PyTorch) and
+                prevents certain floating-point arithmetic-based attacks.
+                See :meth:`~opacus.optimizers.optimizer._generate_noise` for details
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:80]: Also mention this has a dependency on torchcsprng in the docstring itself.

File: opacus/privacy_engine.py [hunk 21]
```diff
@@ -206,7 +260,64 @@ def make_private(
         poisson_sampling: bool = True,
         clipping: str = "flat",
         noise_generator=None,
-    ):
+    ) -> Tuple[GradSampleModule, DPOptimizer, DataLoader]:
+        """
+        Add privacy-related responsibilites to the main PyTorch training objects:
+        model, optimizer and the data loader.
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:266]:
```suggestion
        model, optimizer, and the data loader.
```

File: opacus/privacy_engine.py [hunk 22]
```diff
@@ -206,7 +260,64 @@ def make_private(
...
+
+        All of the returned objects act just like their non-private counterparts
+        passed as arguments, but with added DP tasks.
+
+        Model is wrapped to also compute per sample gradients.
+        Optimizer is now responsible for gradient clipping and adding noise to the
+            gradients
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:273]:
```suggestion
            gradients.
```

File: opacus/privacy_engine.py [hunk 23]
```diff
@@ -206,7 +260,64 @@ def make_private(
...
+        DataLoader is updated to perform Poisson sampling.
+
+        Notes:
+            Using any other models, optimizers or data sources during training
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:277]:
```suggestion
            Using any other models, optimizers, or data sources during training
```

File: opacus/privacy_engine.py [hunk 24]
```diff
@@ -206,7 +260,64 @@ def make_private(
...
+            noise_generator: torch.Generator() object used as a source of randomness for
+                the noise
+
+        Returns:
+            Tuple of (model, optimizer, data_loader).
+
+            Model is a wrapper around the original model that also computes per sample
+                gradients
+            Optimizer is a wrapper around the original optimizer that also does
+             gradient clipping and adding noise to the gradients
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:316]:
```suggestion
             gradient clipping and noise addition to the gradients
```

**Review by karthikprasad — COMMENTED** [review]
LGTM barring a few nits.

File: opacus/privacy_engine.py (hunk 21 above)
Comment (ffuuugor) [reply opacus/privacy_engine.py:266]: much oxford

**Review by karthikprasad — APPROVED** [review]

File: opacus/privacy_engine.py [hunk 25]
```diff
@@ -258,21 +260,23 @@ def make_private(
...
 
         Model is wrapped to also compute per sample gradients.
         Optimizer is now responsible for gradient clipping and adding noise to the
-            gradients
+            gradients.
         DataLoader is updated to perform Poisson sampling.
 
-        Note, that using any other models, optimizers or data sources during training
-        will invalidate stated privacy guarantees.
+        Notes:
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:277]: nit: new line is required before this line (you can just merge this PR and worry about this later)

File: opacus/privacy_engine.py [hunk 26]
```diff
@@ -300,18 +305,21 @@ def make_private(
...
                 per layer clipping in combination with distributed training can provide
                 notable performance gains.
+            noise_generator: torch.Generator() object used as a source of randomness for
+                the noise
 
         Returns:
-            Tuple of (model, optimizer, data_loader)
+            Tuple of (model, optimizer, data_loader).
+
             Model is a wrapper around the original model that also computes per sample
```
Comment (ashkan-software) [inline opacus/privacy_engine.py:314]: I am actually curious to see how this will render on the website. As far as I am concerned, under Returns, only one paragraph can be written (i.e. no new line).

**Review by ashkan-software — APPROVED** [review]
Awesome!
